gpytorch/lazy/lazy_tensor.py

```diff
@@ -732,6 +762,230 @@ def add_jitter(self, jitter_val=1e-3):
+        C = ((A; B); B^T, D). where A is the existing lazy tensor, and B (cross_mat) and D (new_mat)
+        are new components. This is most commonly used when fantasizing with kernel matrices.
+
+        We have access to K \\approx LL^T and K^{-1} \\approx RR^T, where L and R are low rank matrices
+        resulting from root and root inverse decompositions (see the LOVE paper).
+
+        To update R, we first update L:
+            [K U; U^T S] = [L 0; A B][L^T A^T; 0 BT]
+        Solving this matrix equation, we get:
+            K = LL^T ==>       L = L
+            U = LA^T ==>       A = UR^{-1}
+            S = AA^T + BB^T ==> B = cholesky(S - AA^T)
+
+        Once we've computed Z = [L 0; A B], we have that the new kernel matrix [K U; U^T S] \approx ZZ^T. Therefore,
+        we can form a pseudo-inverse of Z directly to approximate [K U; U^T S]^{-1/2}.
+
+        This strategy is also described in "Efficient Nonmyopic Bayesian Optimization via One-Shot Multistep Trees,"
+        Jiang et al, NeurIPS, 2020. https://arxiv.org/abs/2006.15779.
+
+        Args:
+            cross_mat (:obj:`torch.tensor`): the matrix :math:`B` we are appending to the matrix :math:`A`.
+                If :math:`A` is n x n, then this matrix should be n x k.
+            new_mat (:obj:`torch.tensor`): the matrix :math:`D` we are appending to the matrix :math:`A`.
+                If :math:`B` is n x k, then this matrix should be k x k.
+            generate_roots (:obj:`bool`): whether to generate the root decomposition of :math:`A` even if it
+                has not been created yet.
+
+        Returns:
+            :obj:`LazyTensor`: concatenated lazy tensor with the new rows and columns.
+        """
+        from . import lazify
+        from .batch_repeat_lazy_tensor import BatchRepeatLazyTensor
+        from .cat_lazy_tensor import CatLazyTensor
+        from .root_lazy_tensor import RootLazyTensor
+        from .triangular_lazy_tensor import TriangularLazyTensor
+
+        cross_mat = lazify(cross_mat)
+        batch_shape = cross_mat.shape[:-2]
+        new_mat = lazify(new_mat)
+        old_tsr = self
+        if old_tsr.ndimension() < cross_mat.ndimension():
+            old_tsr = BatchRepeatLazyTensor(old_tsr, batch_shape)
+
+        new_lazy_tensor_1 = CatLazyTensor(old_tsr, cross_mat, dim=-2)
+        new_lazy_tensor_2 = CatLazyTensor(cross_mat.transpose(-1, -2), new_mat, dim=-2)
+        new_lazy_tensor = CatLazyTensor(new_lazy_tensor_1, new_lazy_tensor_2, dim=-1)
+
+        # if the old lazy tensor does not have either a root decomposition or a root inverse decomposition
+        # don't create one
+        does_not_have_roots = _is_in_cache_ignore_args(self, "root_decomposition") or _is_in_cache_ignore_args(
+            self, "root_inv_decomposition"
+        )
+        if not generate_roots and not does_not_have_roots:
+            return new_lazy_tensor
+
+        L = self.root_decomposition().root
```

Comment:
You're calling `L = delazify(L)` below anyway, might as well `evaluate` once here...

```suggestion
        L = self.root_decomposition().root.evaluate()
```

Reply:
This is in there because I need to pluck out if L was a cholesky when lazy. let me move some stuff around.